edge-modules/metrics-collector/src/ModuleClientWrapper/IotMessageModuleClientWrapper.cs

```diff
@@ -0,0 +1,60 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Devices.Client;
+using Microsoft.Azure.Devices.Edge.Util;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.Devices.Edge.Azure.Monitor.ModuleClientWrapper
+{
+    public class IotMessageModuleClientWrapper : IDisposable, IModuleClientWrapper
+    {
+        BasicModuleClientWrapper inner;
+        CancellationTokenSource cancellationTokenSource;
+
+        public IotMessageModuleClientWrapper(BasicModuleClientWrapper basicModuleClientWrapper, CancellationTokenSource cts)
+        {
+            this.inner = basicModuleClientWrapper;
+            this.cancellationTokenSource = cts;
+        }
+
+        public static async Task<IModuleClientWrapper> BuildModuleClientWrapperAsync(CancellationTokenSource cts)
+        {
+            BasicModuleClientWrapper basicModuleClientWrapper = await BasicModuleClientWrapper.BuildModuleClientWrapperAsync();
+            return new IotMessageModuleClientWrapper(basicModuleClientWrapper, cts);
+        }
+
+        public async Task RecreateClientAsync()
+        {
+            try
+            {
+                await this.inner.RecreateClientAsync();
+            }
+            catch (Exception e)
+            {
+                LoggerUtil.Writer.LogError($"Failed closing and re-establishing connection to IoT Hub: {e.ToString()}");
+                this.cancellationTokenSource.Cancel();
```

Comment:
This cancellation token is waited on inside main() to stop the program. In the IoTMessage case, if we can't close and re-establish connection we will fail fast.

Reply:
I understand why it's being cancelled here. I don't understand why the token isn't being passed into other async methods in the code path. For example, shouldn't BasicModuleClientWrapper.BuildModuleClientWrapperAsync() take a token parameter, which can be passed into ModuleClient.CreateFromEnvironmentAsync() and ModuleClient.OpenAsync()? Likewise, shouldn't BasicModuleClientWrapper.RecreateClientAsync() and BasicModuleClientWrapper.SendMessageAsync() take token parameters?

Reply:
Thinking about it some more...we could do more work here (and throughout the codebase) to make more effective use of cancellation tokens, but what you have here does what it was intended to do (cancels module scraping and exits the module if client refresh fails, allowing the module to restart and create a new client instead of waiting up to one day). So let's stick with this.